kas/kdl/sema/define_directive.cpp

```diff
@@ -0,0 +1,325 @@
+    
+    if (sema->expect({ kdl::condition(kdl::lexer::token::type::string).falsey() })) {
+        log::error(sema->peek().file(), sema->peek().line(), "Type definition field name should be a string.");
+    }
+    auto field_name = sema->read().text();
+    
+    sema->ensure({ kdl::condition(kdl::lexer::token::type::rparen).truthy() });
+    
+    return field_name;
+}
+
+static inline kdk::assembler::field::value parse_field_value(kdl::sema *sema)
+{
+    auto file = sema->peek().file();
+    auto line = sema->peek().line();
+    
+    sema->ensure({ kdl::condition(kdl::lexer::token::type::lparen).truthy() });
+    
+    // Default attributes - all attributes are provided as strings ready to be parsed
+    // later
+    kdk::assembler::field::value::type value_type { kdk::assembler::field::value::type::integer };
+    uint64_t value_size { 0 };
+    uint64_t value_length { 0 };
+    std::string value_name;
+    uint64_t value_offset { 0 };
+    
+    bool length_required = false;
+    bool size_required = false;
+    
+    while (sema->expect({ kdl::condition(kdl::lexer::token::type::rparen).falsey() })) {
+        
+        if (sema->expect({ kdl::condition(kdl::lexer::token::type::identifier).falsey(), kdl::condition(kdl::lexer::token::type::equals).falsey() })) {
+            log::error(sema->peek().file(), sema->peek().line(), "Malformed value attribute encountered.");
+        }
+        auto attribute = sema->read().text();
+        sema->advance();
+        
+        if (attribute == "name") {
+            if (sema->expect({ kdl::condition(kdl::lexer::token::type::string).falsey() })) {
+                log::error(sema->peek().file(), sema->peek().line(), "The name attribute of a type definition value must be a string.");
+            }
+            value_name = sema->read().text();
+        }
+        else if (attribute == "offset") {
+            if (sema->expect({ kdl::condition(kdl::lexer::token::type::integer).falsey() })) {
+                log::error(sema->peek().file(), sema->peek().line(), "The offset attribute of a type definition value must be an integer.");
+            }
+            value_offset = std::stoull(sema->read().text());
+        }
+        else if (attribute == "length") {
+            if (sema->expect({ kdl::condition(kdl::lexer::token::type::integer).falsey() })) {
+                log::error(sema->peek().file(), sema->peek().line(), "The length attribute of a type definition value must be an integer.");
+            }
+            value_length = std::stoull(sema->read().text());
+        }
+        else if (attribute == "size") {
+            if (sema->expect({ kdl::condition(kdl::lexer::token::type::integer).truthy() })) {
+                value_size = std::stoull(sema->read().text());
+            }
+            else if (sema->expect({ kdl::condition(kdl::lexer::token::type::identifier).truthy() })) {
+                auto size_symbol = sema->read().text();
```

Comment:
Note to self. The code to parse a size symbol could be pulled to its own function.